Feature/nickname

JungMinji08 opened the pull request:
- 닉네임 생성 화면 추가
- Retrofit + Moshi 네트워크 모듈 추가
- https://api-dev.damgle.com/v1/namepicker API를 통해 랜덤 닉네임 생성

JungMinji08 committed:
Add Network(Retrofit + Moshi) Module & Add GetRandomNickName logic

- 네트워크 모듈 생성(Retrofit + Moshi)
- 랜덤 닉네임 불러오는 로직 추가 (실제 네트워크 통신)

JungMinji08 committed:
Add Refresh NickName Function(Adjective, Noun)

- 형용사, 명사 각각 새로고침 할 수 있도록 usecase 수정
- 클릭이벤트 발생 시 새로 받아올 데이터를 임시로 blank 처리해서 함수를 호출하면, blank 되지 않는 옵션을 서버로 보내서 새로운 데이터를 받아옴

JungMinji08 committed:
Add ViewState sealed class

- ViewState 추가 (Success, Error, Loading)

JungMinji08 committed:
Add NickNameScreen, NickNameBox(Composable) UI & Add OnboardingViewModel

- 닉네임 스크린 및 닉네임 컴포저블 함수 추가 UI 추가
- OnboardingViewModel 추가
- ic_refresh 추가
- 닉네임 페이지에서 사용될 strings 추가

JungMinji08 commented:
네트워크 모듈 추가했는데 PR이 너무 커지는거같아서 우선 header에 토큰 넣는거나 BaseRequest, BaseResponse 이런건 필요할 때 추가로 작성하도록 해도 괜찮을까요?

JungMinji08 commented on data/src/main/java/com/mashup/damgledamgle/repository/spec/NickNameRequest.kt:
```diff
@@ -0,0 +1,17 @@
+ *  Copyright © 2022 MashUp All rights reserved.
+ */
+
+data class NickNameRequest(
```
아 이거 Response로 바꿔야 한다...;;

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/NickNameScreen.kt:
```diff
@@ -4,64 +4,164 @@ import androidx.compose.foundation.Image
+ *  Copyright © 2022 MashUp All rights reserved.
  */
 
 @Composable
```
요거 따로 패키지 빼줘도 될듯? 하네요
큰 상관은 없습니당

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/NickNameScreen.kt:
```diff
@@ -4,64 +4,164 @@ import androidx.compose.foundation.Image
+                        modifier = Modifier.padding(horizontal = 20.dp, vertical = 16.dp)
+                    ) {
+                        NickNameBox(viewModel.nickName.value.adjective) {
+                            viewModel.refreshNickName(
+                                viewModel.nickName.value.copy(
+                                    adjective = ""
+                                )
+                            )
+                        }
+
+                        NickNameBox(viewModel.nickName.value.noun) {
+                            viewModel.refreshNickName(
+                                viewModel.nickName.value.copy(
+                                    noun = ""
+                                )
```
refresh 함수 두 개로 분리하는게 더 깔끔할거같아요
(내부적으로 refreshNickName쓰더라두?)

dahyun1226 commented on app/src/main/java/com/mashup/damgledamgle/presentation/feature/onboarding/NickNameScreen.kt:
```diff
@@ -4,64 +4,164 @@ import androidx.compose.foundation.Image
+@Preview
+@Composable
+fun PreviewNickNameBox() {
+    NickNameBox("닉네임")
```
리프레시가 약간 아래인듯하네요

dahyun1226 commented on data/src/main/java/com/mashup/damgledamgle/repository/remote/OnboardingRepositoryImpl.kt:
```diff
@@ -0,0 +1,35 @@
+    private val nickNameMapper: NickNameMapper
+) : OnboardingRepository {
+    override suspend fun getNickName(adjective: String?, noun: String?): NetworkResponse<NickName> {
+        val resultData = if (adjective != null) {
+            ServiceBuilder.damgleApi.getNickName(adjective = adjective)
+        } else if (noun != null) {
+            ServiceBuilder.damgleApi.getNickName(noun = noun)
+        } else {
+            ServiceBuilder.damgleApi.getNickName()
+        }
```
```suggestion
                val resultData = when {
            adjective != null -> ServiceBuilder.damgleApi.getNickName(adjective = adjective)
            noun != null -> ServiceBuilder.damgleApi.getNickName(noun = noun)
            else -> ServiceBuilder.damgleApi.getNickName()
        }
```
개인적으로 패턴매칭을 굉장히 좋아합니다! 더 깔끔하게 느껴져요

dahyun1226 reviewed (changes requested):
1. 아이디 랜덤으로 바뀌면서 같은거 걸리면 다시 돌려주던가 해야할듯? 돌리는 의미가 없으니까.
아니면 api 자체를 좀 바꿔달라고 요청하던지!

2. 이름 변하면 밑줄도 사이즈 변하는게 기획적으로 맞는건지? (약간 어색함이 있네요)

3. 네트워크 끄면 앱 터집니닷. 
android.util.MalformedJsonException 뜨는거 같아요 (맞는지 모름). 수정부탁드립니다!

JungMinji08 committed:
Fix NickNameResponse Naming from NickNameRequest

JungMinji08 committed:
Add SignUp Api

JungMinji08 committed:
Refactor Code

- 코드리뷰 반영
- 닉네임 리프레쉬 함수 (형용사, 명사) 분리
- if else 패턴 when으로 변경

JungMinji08 committed:
Add Token to header when post api

- API 통신 보낼 때 헤더에 토큰 넣어서 보냄

JungMinji08 committed:
Fix Refresh Icon Position

- 리프레쉬 아이콘 위치 수정

JungMinji08 commented:
리뷰 반영했습니다

그리고...

1. 아이디 랜덤으로 바뀌면서 같은거 걸리면 다시 돌려주던가 해야할듯? 돌리는 의미가 없으니까.
아니면 api 자체를 좀 바꿔달라고 요청하던지!
====> 나중에 서버쪽에 물어보고 변경해달라고 합시다! ㅎㅎ. 그리고 지금 중복 아이디도 내려오고 있는거같은데 이 부분도 말해봐야할거같아여

2. 이름 변하면 밑줄도 사이즈 변하는게 기획적으로 맞는건지? (약간 어색함이 있네요)
====> 내일 QA하면서 물어보고 UI 수정하겠습니당

네트워크 끄면 앱 터집니닷.
android.util.MalformedJsonException 뜨는거 같아요 (맞는지 모름). 수정부탁드립니다!
====> 네트워크 통신 보내는 부분부터 try/catch로 잡아서 처리했습니다. 나중에 NetworkResponse.Error면 다이얼로그 띄우는거 추가할게여

JungMinji08 committed:
Add navigation logic when SignUp Succeed

JungMinji08 committed:
Add `Save Refresh Token Logic` & Fix SignUp Request Model

- 리프레쉬 토큰 저장하는 로직 추가
- SignUp Request Model 추가
- `invoke()` 제거
- `value` -> `emit()`으로 교체

JungMinji08 committed:
Fix hard coding
